File: problems/2026/04/2026-04-01-leetcode-5-最长回文子串/SolutionLC5.java
/**
 * LeetCode 5：最长回文子串（中心扩展，O(n^2)）
 */
public class SolutionLC5 {

    public String longestPalindrome(String s) {
        if (s == null || s.isEmpty()) {
            return "";
        }
        // 最长回文在 s 中的起始下标与长度
        int start = 0;
        int maxLen = 0;

        for (int i = 0; i < s.length(); i++) {
            // 每个下标既可能是奇数长度回文的中心，也可能是偶数回文中心的左端点
            // 奇数长度：中心为一个字符
            int len1 = expandAroundCenter(s, i, i);
            // 偶数长度：中心为两个字符之间
            int len2 = expandAroundCenter(s, i, i + 1);
            int len = Math.max(len1, len2);
            if (len > maxLen) {
                maxLen = len;
                // 回文左端点：中心向左偏移 (len-1)/2
                start = i - (len - 1) / 2;
            }
        }
        return s.substring(start, start + maxLen);
    }

    /**
     * 从 [left, right] 向两侧扩展，返回回文长度（至少为 right-left+1）。
     */
    private int expandAroundCenter(String s, int left, int right) {
        while (left >= 0 && right < s.length() && s.charAt(left) == s.charAt(right)) {
            left--;
            right++;
        }
        // 循环退出时多扩了一步，长度为 right - left - 1
        return right - left - 1;
    }

    private static boolean isPalindrome(String t) {
        int l = 0;
        int r = t.length() - 1;
        while (l < r) {
            if (t.charAt(l++) != t.charAt(r--)) {
                return false;
            }
        }
        return true;
    }

    public static void main(String[] args) {
        SolutionLC5 solution = new SolutionLC5();

        String[] inputs = {"babad", "cbbd", "a", "ac", ""};
        int[] expectedLens = {3, 2, 1, 1, 0};
        String[] expectedDesc = {
                "bab 或 aba",
                "bb",
                "a",
                "a 或 c",
                "（空串）"
        };

        for (int i = 0; i < inputs.length; i++) {
            String in = inputs[i];
            String result = solution.longestPalindrome(in);
            boolean ok = result.length() == expectedLens[i]
                    && (expectedLens[i] == 0 ? result.isEmpty() : in.contains(result))
                    && isPalindrome(result);
            System.out.println("调用: longestPalindrome(\"" + in + "\")");
            System.out.println("预期结果: " + expectedDesc[i] + "（长度 " + expectedLens[i] + "）");
            System.out.println("实际结果: " + (result.isEmpty() ? "（空串）" : result));
            System.out.println("校验: " + (ok ? "OK" : "FAIL"));
            System.out.println();
        }
    }
}

File: problems/2026/04/2026-04-01-leetcode-5-最长回文子串/SolutionLC5.py
class Solution:
    def longest_palindrome(self, s: str) -> str:
        """中心扩展：O(n^2) 时间，O(1) 额外空间（不计返回串）。"""
        if not s:
            return ""

        def expand(left: int, right: int) -> int:
            # 向两侧扩展，退出时多走一步，长度为 right - left - 1
            while left >= 0 and right < len(s) and s[left] == s[right]:
                left -= 1
                right += 1
            return right - left - 1

        start = 0
        max_len = 0
        for i in range(len(s)):
            # 奇数中心 (i,i)、偶数中心 (i,i+1) 各扩一次
            len1 = expand(i, i)
            len2 = expand(i, i + 1)
            cur = max(len1, len2)
            if cur > max_len:
                max_len = cur
                start = i - (cur - 1) // 2
        return s[start : start + max_len]


def _is_palindrome(t: str) -> bool:
    return t == t[::-1]


if __name__ == "__main__":
    solution = Solution()

    inputs = ["babad", "cbbd", "a", "ac", ""]
    expected_lens = [3, 2, 1, 1, 0]
    expected_desc = ["bab 或 aba", "bb", "a", "a 或 c", "（空串）"]

    for i, inp in enumerate(inputs):
        result = solution.longest_palindrome(inp)
        ok = (
            len(result) == expected_lens[i]
            and (expected_lens[i] == 0 or result in inp)
            and _is_palindrome(result)
        )
        print(f'调用: longest_palindrome("{inp}")')
        print(f"预期结果: {expected_desc[i]}（长度 {expected_lens[i]}）")
        display = "（空串）" if result == "" else result
        print(f"实际结果: {display}")
        print(f"校验: {'OK' if ok else 'FAIL'}")
        print()
